### ISSUE TITLE
This training is useless

### ISSUE THREAD

**Issue by cyilu** [tensorflow/tfjs-core#879]
There is my training code:
`

    function train(myData) {
       const returnCost = true;
       for (let i = 0; i < TRAIN_STEPS; i++) {
            dl.tidy(() => {
            const cost = optimizer.minimize(() => {

                const batch = myData.nextTrainBatch(BATCH_SIZE);
                let x_image = dl.reshape(batch.xs, [-1, 28, 28, 1]);
                let b = dl.add(conv2d(x_image, W_conv1), b_conv1);
                let r_conv1 = dl.relu(b);
                let r_pool1 = max_pool_2x2(r_conv1);
                let r_conv2 = dl.relu(dl.add(conv2d(r_pool1, W_conv2), b_conv2));
                let r_pool2 = max_pool_2x2(r_conv2);
                let r_pool2_flat = dl.reshape(r_pool2, [-1, 7 * 7 * 64]);
                let r_fc1 = dl.relu(dl.add(dl.matMul(r_pool2_flat, W_fc1), b_fc1));
                let y_pre = dl.softmax(dl.add(dl.matMul(r_fc1, W_fc2), b_fc2));
                return myMath.mean(dl.losses.softmaxCrossEntropy(batch.labels, y_pre));

            }, returnCost);
            if (i % 10 == 0) {
                console.log(i);
                cost.print();
            }
        });
    }
}

`

However, the "cost" don't decrease.

**Comment by dsmilkov** [comment]
Quick q: What is `myMath` referring to? Also can you share with us your full code/repo so we get better investigate?

**Comment by cyilu** [comment]
My full code is:
`
import * as dl from 'deeplearn';
const TRAIN_TEST_RATIO = 5 / 6;
const LEARNING_RATE = 0.2;
const BATCH_SIZE = 64;
const IMAGE_SIZE = 784;
const LABELS_SIZE = 10;
const TRAIN_STEPS = 100;

const myMath = dl.ENV.math;

function weight_variable(shape) {
    let init = dl.truncatedNormal(shape, 0, 0.1, "float32");
    return dl.variable(init);
}

function bias_variable(shape) {
    let init = dl.fill(shape, 0.1);
    return dl.variable(init);
}

function conv2d(x, W) {
    return dl.conv2d(x, W, 1, "same");
}

function max_pool_2x2(x) {
    return dl.maxPool(x, 2, 2, "valid");
}

let W_conv1 = weight_variable([5, 5, 1, 32]);
let b_conv1 = bias_variable([32]);
let W_conv2 = weight_variable([5, 5, 32, 64]);
let b_conv2 = bias_variable([64]);

let W_fc1 = weight_variable([7 * 7 * 64, 1024]);
let b_fc1 = bias_variable([1024]);

let W_fc2 = weight_variable([1024, 10]);
let b_fc2 = bias_variable([10]);

const optimizer = dl.train.sgd(LEARNING_RATE);

const mnistConfig = {
    'data': [
        {
            'name': 'images',
            'path': 'https://storage.googleapis.com/learnjs-data/model-builder/' +
                'mnist_images.png',
            'dataType': 'png',
            'shape': [28, 28, 1]
        },
        {
            'name': 'labels',
            'path': 'https://storage.googleapis.com/learnjs-data/model-builder/' +
                'mnist_labels_uint8',
            'dataType': 'uint8',
            'shape': [10]
        }
    ],
    modelConfigs: {}
};

class MnistData {

    constructor() {
        this.shuffledTrainIndex = 0;
        this.shuffledTestIndex = 0;
    }


    nextTrainBatch(batchSize) {
        return this.nextBatch(batchSize, this.trainingData, () => {
            this.shuffledTrainIndex = (this.shuffledTrainIndex + 1) % this.trainIndices.length;
            return this.trainIndices[this.shuffledTrainIndex];
        });
    }



    nextTestBatch(batchSize) {
        return this.nextBatch(batchSize, this.testData, () => {
            this.shuffledTestIndex =
                (this.shuffledTestIndex + 1) % this.testIndices.length;
            return this.testIndices[this.shuffledTestIndex];
        });
    }



    nextBatch(batchSize, myData, index) {
        let xs = null;
        let labels = null;
        for (let i = 0; i < batchSize; i++) {
            const idx = index();
            const x = myData[0][idx].reshape([1, 784]);
            xs = concatWithNulls(xs, x);
            const label = myData[1][idx].reshape([1, 10]);
            labels = concatWithNulls(labels, label);
        }
        return { xs, labels };
    }

    // Load all the MNIST data 
    async load() {
        this.dataset = new dl.XhrDataset(mnistConfig);
        await this.dataset.fetchData();
        this.dataset.normalizeWithinBounds(0, -1, 1);
        this.trainingData = this.getTrainingData();
        this.testData = this.getTestData();
        this.trainIndices = dl.util.createShuffledIndices(this.trainingData[0].length);
        this.testIndices = dl.util.createShuffledIndices(this.testData[0].length);
    }


    getTrainingData() {
        const [images, labels] = this.dataset.getData();
        const end = Math.floor(TRAIN_TEST_RATIO * images.length);
        return [images.slice(0, end), labels.slice(0, end)];
    }


    getTestData() {
        const myData = this.dataset.getData();
        if (myData == null) { return null; }
        const [images, labels] = this.dataset.getData();
        const start = Math.floor(TRAIN_TEST_RATIO * images.length);
        return [images.slice(start), labels.slice(start)];
    }

}  // end of class MnistData  definition

// A helper for the NextBatch function above
function concatWithNulls(ndarray1, ndarray2) {
    if (ndarray1 == null && ndarray2 == null) {
        return null;
    }
    if (ndarray1 == null) {
        return ndarray2;
    }
    else if (ndarray2 === null) {
        return ndarray1;
    }
    return myMath.concat2D(ndarray1, ndarray2, 0);
}

// Train the model.
function train(myData) {
    const returnCost = true;
    for (let i = 0; i < TRAIN_STEPS; i++) {
        const cost = optimizer.minimize(() => {
            const batch = myData.nextTrainBatch(BATCH_SIZE);
            return loss(batch);
        }, returnCost);
        if (i % 10 == 0) {
            console.log(i);
            cost.print();
        };
    }
}

function loss(batch) {
    let y_pre = predict(batch);
    return dl.mean(dl.losses.softmaxCrossEntropy(batch.labels, y_pre));
}

function predict(batch) {
    return dl.tidy(() => {
        let x_image = dl.reshape(batch.xs, [-1, 28, 28, 1]);
        let b = dl.add(conv2d(x_image, W_conv1), b_conv1);
        let r_conv1 = dl.relu(b);
        let r_pool1 = max_pool_2x2(r_conv1);
        let r_conv2 = dl.relu(dl.add(conv2d(r_pool1, W_conv2), b_conv2));
        let r_pool2 = max_pool_2x2(r_conv2);
        let r_pool2_flat = dl.reshape(r_pool2, [-1, 7 * 7 * 64]);
        let r_fc1 = dl.relu(dl.add(dl.matMul(r_pool2_flat, W_fc1), b_fc1));
        let y_pre = dl.softmax(dl.add(dl.matMul(r_fc1, W_fc2), b_fc2));
        return y_pre;
    });
}

// Main Entry point
async function myMnist() {

    // Load Mnist Dataset
    console.log("Loading...");
    let myData = new MnistData();
    await myData.load();

    // Train the network
    console.log("Training...");
    await train(myData);

    //Predict
    const testExamples = 100;
    const batch = myData.nextTestBatch(testExamples);
    const predictions = predict(batch);
    let correct = 0;
    const axis = 1;

    const y_corret = dl.argMax(batch.labels, axis);

    const y_pred = dl.argMax(predictions, axis);

    let r_equ=dl.equalStrict(y_corret, y_pred);
    let s=dl.sum(r_equ);
    let acc=dl.div(s,dl.tensor1d([testExamples]) );
    s.print();
    acc.print();
}  // End of Main program

myMnist();`

I don't konw where I am wrong. Please help me.

**Comment by dsmilkov** [comment]
Hi,

Two small finds:
- dl.softmaxCrossEntropy expects the pre-softmax *logits* as a 2nd argument, so you have to remove the `softmax` call in your predict function.
- Lower the learning rate to 0.01/0.03

This should help with the training.

**Comment by cyilu** [comment]
Thank you so much！！ It is right！！  But why my computing time is  a little  long, but the "mnist eager" demo run fastly. And the browser is slow.
